source/extensions/upstreams/http/udp/upstream_request.cc

```diff
@@ -0,0 +1,141 @@
+                                               bool /*end_stream*/) {
+  Api::SysCallIntResult rc = socket_->connect(host_->address());
+  if (SOCKET_FAILURE(rc.return_value_)) {
+    return absl::InternalError("Upstream socket connect failure.");
+  }
+  // Synthesize the 200 response headers downstream to complete the CONNECT-UDP handshake.
+  Envoy::Http::ResponseHeaderMapPtr headers{
+      Envoy::Http::createHeaderMap<Envoy::Http::ResponseHeaderMapImpl>(
+          {{Envoy::Http::Headers::get().Status, "200"},
+           {Envoy::Http::Headers::get().CapsuleProtocol, "?1"}})};
+  upstream_to_downstream_->decodeHeaders(std::move(headers), false);
+  return Envoy::Http::okStatus();
+}
+
+void UdpUpstream::resetStream() { upstream_to_downstream_ = nullptr; }
+
+void UdpUpstream::onSocketReadReady() {
+  uint32_t packets_dropped = 0;
+  const Api::IoErrorPtr result = Network::Utility::readPacketsFromSocket(
+      socket_->ioHandle(), *socket_->connectionInfoProvider().localAddress(), *this,
+      dispatcher_.timeSource(), /*prefer_gro=*/false, packets_dropped);
+  if (result == nullptr) {
+    socket_->ioHandle().activateFileEvents(Event::FileReadyType::Read);
+    return;
+  }
+}
+
+void UdpUpstream::processPacket(Network::Address::InstanceConstSharedPtr /*local_address*/,
+                                Network::Address::InstanceConstSharedPtr /*peer_address*/,
+                                Buffer::InstancePtr buffer, MonotonicTime /*receive_time*/) {
+  std::string data = buffer->toString();
+  quiche::QuicheBuffer serialized_capsule =
+      SerializeCapsule(quiche::Capsule::Datagram(data), &capsule_buffer_allocator_);
+
+  Buffer::InstancePtr capsule_data = std::make_unique<Buffer::OwnedImpl>();
+  capsule_data->add(serialized_capsule.AsStringView());
+  bytes_meter_->addWireBytesReceived(capsule_data->length());
+  upstream_to_downstream_->decodeData(*capsule_data, false);
+}
+
+bool UdpUpstream::OnCapsule(const quiche::Capsule& capsule) {
+  quiche::CapsuleType capsule_type = capsule.capsule_type();
+  if (capsule_type != quiche::CapsuleType::DATAGRAM) {
+    // Silently drops Datagram Capsules with an unknown type.
+    return true;
+  }
+
+  std::unique_ptr<quiche::ConnectUdpDatagramPayload> connect_udp_datagram_payload =
+      quiche::ConnectUdpDatagramPayload::Parse(capsule.datagram_capsule().http_datagram_payload);
+  if (!connect_udp_datagram_payload) {
+    // Indicates parsing failure to reset the data stream.
+    return false;
+  }
+
+  if (connect_udp_datagram_payload->GetType() !=
+      quiche::ConnectUdpDatagramPayload::Type::kUdpPacket) {
+    // Silently drops Datagrams with an unknown Context ID.
+    return true;
+  }
+
+  Buffer::InstancePtr buffer = std::make_unique<Buffer::OwnedImpl>();
+  buffer->add(connect_udp_datagram_payload->GetUdpProxyingPayload());
+  bytes_meter_->addWireBytesSent(buffer->length());
+  Api::IoCallUint64Result rc = Network::Utility::writeToSocket(
```

Comment:
Are we ignoring socket write errors? If so, can we add a TODO to react to some errors? For example if we get "no route to host" we could close the stream. Similarly, I notice that this file tracks the number of dropped datagrams, we could increase that number when we get socket write errors

Reply:
Good point. I added a TODO. The existing CONNECT implementation doesn't keep track of errors so perhaps I could enhance the CONNECT implementation together.

Reply:
SGTM